core/services/gateway/handlers/common/message_util.go

```diff
@@ -0,0 +1,97 @@
+package common
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	jsonrpc "github.com/smartcontractkit/chainlink-common/pkg/jsonrpc2"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/api"
+)
+
+func ValidatedMessageFromResp(resp *jsonrpc.Response) (*api.Message, error) {
+	if resp.Error != nil {
+		return nil, fmt.Errorf("received error, ID: %s", resp.ID)
+	}
+	if resp.Result == nil {
+		return nil, fmt.Errorf("response result is nil, ID: %s", resp.ID)
+	}
+	var msg api.Message
+	err := json.Unmarshal(resp.Result, &msg)
+	if err != nil {
+		return nil, err
+	}
+	msg.Body.MessageId = resp.ID
+	err = msg.Validate()
+	if err != nil {
+		return nil, err
+	}
+	return &msg, nil
+}
+
+func ValidatedMessageFromReq(req *jsonrpc.Request) (*api.Message, error) {
```

Comment:
Nice! Can you add a comment explaining that this essentially converts a generic parsed JSON RPC message into the "legacy" Message struct?

Reply:
done 👍